src/main/scala/is/hail/annotations/UnsafeRow.scala

```diff
@@ -1,134 +1,291 @@
+  def subtree(i: Int): BroadcastTypeTree = subtrees(i)
+}
+
+class UnsafeIndexedSeqAnnotation(region: MemoryBuffer,
+  arrayTTBc: BroadcastTypeTree,
+  elemSize: Int, offset: Int, elemOffset: Int,
+  val length: Int) extends IndexedSeq[Annotation] {
+  def apply(i: Int): Annotation = {
+    assert(i >= 0 && i < length)
+    if (region.loadBit(offset + 4, i))
+      null
+    else
+      UnsafeRow.read(region, elemOffset + i * elemSize,
+        arrayTTBc.value.typ.asInstanceOf[TContainer].elementType,
+        arrayTTBc.value.subtree(0))
+  }
+}
 
-  private def readBinary(offset: Int): Array[Byte] = {
-    val start = mb.loadInt(offset)
+object UnsafeRow {
+  def readBinary(region: MemoryBuffer, offset: Int): Array[Byte] = {
+    val start = region.loadInt(offset)
     assert(offset > 0 && (offset & 0x3) == 0, s"invalid binary start: $offset")
-    val binLength = mb.loadInt(start)
-    val b = mb.loadBytes(start + 4, binLength)
+    val binLength = region.loadInt(start)
+    val b = region.loadBytes(start + 4, binLength)
 
     b
   }
 
-  private def readArray(offset: Int, t: Type): IndexedSeq[Any] = {
-    val start = mb.loadInt(offset)
+  def readArray(region: MemoryBuffer, offset: Int, elemType: Type, arrayTTBc: BroadcastTypeTree): IndexedSeq[Any] = {
+    val aoff = region.loadInt(offset)
 
-    assert(start > 0 && (start & 0x3) == 0, s"invalid array start: $offset")
+    val length = region.loadInt(aoff)
+    val elemOffset = UnsafeUtils.roundUpAlignment(aoff + 4 + (length + 7) / 8, elemType.alignment)
+    val elemSize = UnsafeUtils.arrayElementSize(elemType)
 
-    val arrLength = mb.loadInt(start)
-    val missingBytes = (arrLength + 7) / 8
-    val elemsStart = UnsafeUtils.roundUpAlignment(start + 4 + missingBytes, t.alignment)
-    val eltSize = UnsafeUtils.arrayElementSize(t)
+    new UnsafeIndexedSeqAnnotation(region, arrayTTBc, elemSize, aoff, elemOffset, length)
+  }
+
+  def readStruct(region: MemoryBuffer, offset: Int, ttBc: BroadcastTypeTree): UnsafeRow = {
+    new UnsafeRow(ttBc, region, offset)
+  }
 
-    val a = new Array[Any](arrLength)
+  def readString(region: MemoryBuffer, offset: Int): String =
+    new String(readBinary(region, offset))
 
-    var i = 0
-    while (i < arrLength) {
+  def readLocus(region: MemoryBuffer, offset: Int): Locus = {
+    val repr = TLocus.representation
+    Locus(
+      readString(region, offset + repr.byteOffsets(0)),
+      region.loadInt(offset + repr.byteOffsets(1)))
+  }
+
+  def readAltAllele(region: MemoryBuffer, offset: Int): AltAllele = {
+    val repr = TAltAllele.representation
+    AltAllele(
+      readString(region, offset + repr.byteOffsets(0)),
+      readString(region, offset + repr.byteOffsets(1)))
+  }
 
-      val byteIndex = i / 8
-      val bitShift = i & 0x7
-      val missingByte = mb.loadByte(start + 4 + byteIndex)
-      val isMissing = (missingByte & (0x1 << bitShift)) != 0
+  def readArrayAltAllele(region: MemoryBuffer, offset: Int): Array[AltAllele] = {
```

Comment:
I can't because I don't have the right BroadcastTypeTree.  I originally had BroadcastTypeTree contain a broadcast of all the complex standard types, but that was a huge mess.  I think this is cleaner.  (And I also temporary.)